// @jaspr/client-js/lib/Expression.js
export var Constants;
(function (Constants) {
    // LOGICAL
    Constants["LOGICAL_EQUAL"] = "eq";
    Constants["LOGICAL_NOT_EQUAL"] = "ne";
    Constants["LOGICAL_GREATER_THAN"] = "gt";
    Constants["LOGICAL_GREATER_THAN_OR_EQUAL"] = "ge";
    Constants["LOGICAL_LOWER_THAN"] = "lt";
    Constants["LOGICAL_LOWER_THAN_OR_EQUAL"] = "le";
    Constants["LOGICAL_AND"] = "and";
    Constants["LOGICAL_OR"] = "or";
    Constants["LOGICAL_NOT"] = "not";
    Constants["LOGICAL_HAS"] = "has";
    Constants["LOGICAL_IN"] = "in";
    Constants["LOGICAL_BETWEEN"] = "be";
    // ARITHMETIC
    Constants["ARITHMETIC_ADDITION"] = "add";
    Constants["ARITHMETIC_SUBTRACTION"] = "sub";
    Constants["ARITHMETIC_NEGATION"] = "-";
    Constants["ARITHMETIC_MULTIPLICATION"] = "mul";
    Constants["ARITHMETIC_DIVISION"] = "div";
    Constants["ARITHMETIC_MODULO"] = "mod";
    // FUNCTION
    /* String and Collection */
    Constants["FUNCTION_STARTS_WITH"] = "startsWith";
    Constants["FUNCTION_ENDS_WITH"] = "endsWith";
    Constants["FUNCTION_CONTAINS"] = "contains";
    Constants["FUNCTION_CONCAT"] = "concat";
    Constants["FUNCTION_INDEX_OF"] = "indexof";
    Constants["FUNCTION_LENGTH"] = "length";
    Constants["FUNCTION_SUBSTRING"] = "substring";
    /* String */
    Constants["FUNCTION_MATCHES_PATTERN"] = "matchesPattern";
    Constants["FUNCTION_TO_LOWER"] = "tolower";
    Constants["FUNCTION_TO_UPPER"] = "toupper";
    Constants["FUNCTION_TRIM"] = "trim";
    /* Lambda operators */
    Constants["FUNCTION_ANY"] = "any";
    Constants["FUNCTION_ALL"] = "all";
    /* Date & time */
    Constants["FUNCTION_DATE"] = "date";
    Constants["FUNCTION_DAY"] = "day";
    Constants["FUNCTION_HOUR"] = "hour";
    Constants["FUNCTION_MINUTE"] = "minute";
    Constants["FUNCTION_MONTH"] = "month";
    Constants["FUNCTION_NOW"] = "now";
    Constants["FUNCTION_SECOND"] = "second";
    Constants["FUNCTION_TIME"] = "time";
    Constants["FUNCTION_YEAR"] = "year";
    /* Arithmetic */
    Constants["FUNCTION_CEILING"] = "ceiling";
    Constants["FUNCTION_FLOOR"] = "floor";
    Constants["FUNCTION_ROUND"] = "round";
})(Constants || (Constants = {}));
