// @jaspr/client-js/lib/NativeFetchClient.js
export default function (init) {
    const headers = new Headers();
    const reqInit = Object.assign({
        method: 'GET',
        headers: headers
    }, init);
    return {
        delete(url, body, signal) {
            reqInit.signal = signal;
            reqInit.method = 'DELETE';
            reqInit.body = body;
            return fetch(url, reqInit);
        },
        get(url, signal) {
            reqInit.signal = signal;
            reqInit.method = 'GET';
            reqInit.body = undefined;
            return fetch(url, reqInit);
        },
        patch(url, body, signal) {
            reqInit.signal = signal;
            reqInit.method = 'PATCH';
            reqInit.body = body;
            return fetch(url, reqInit);
        },
        post(url, body, signal) {
            reqInit.signal = signal;
            reqInit.method = 'POST';
            reqInit.body = body;
            return fetch(url, reqInit);
        },
        setHeader(name, value) {
            headers.set(name, value);
        }
    };
}
